Derive skip

astrale-sharp opened the pull request:
Adds variant(skip) To/FromVariant proc macro

astrale-sharp commented:
Note that this is based upon my previous PR (which should obviously be merged first)

astrale-sharp commented:
added missing godot::builtin::, sorry about that

Bromeon commented:
There are merge conflicts, could you resolve them please, @astrale-sharp?

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-251

Bromeon commented:
What are your plans with this PR?

CI is currently red, and there is stll the open question how we should _deserialize_ a struct with a skipped field. What value should it get? `Default::default()`?

astrale-sharp commented:
For me it should get Default::default() yes, unless a #[variant(default = ...)] is there.

Originally i was blocked by ./check.sh itest crashing but this seems to be resolved so I'm going to work on this !

astrale-sharp commented:
Hi!

I added functionality to filter test, i still need to update the chech.sh docs but you can now run ./check.sh itest substring, (it would only run test containing substring)

added basic skip attributes using default, next part is support for specifiying the default in your attribute.

Bromeon commented on itest/rust/src/derive_variant/mod.rs:
```diff
@@ -9,7 +9,42 @@ use std::fmt::Debug;
-use godot::builtin::{dict, varray, FromVariant, ToVariant};
+use godot::builtin::{dict, varray, FromVariant, ToVariant, Variant};
+
+#[macro_export]
+macro_rules! testit {
```
Please better name

astrale-sharp replied:
would roundtrip_with_skip or roundtrip_with work?

Bromeon reviewed (commented):
> I added functionality to filter test, i still need to update the chech.sh docs but you can now run ./check.sh itest substring, (it would only run test containing substring)
if you want it in another PR just say so, no problem.

Yes, please don't mix unrelated functionality in the same PR unless it was previously agreed. This PR should ideally focus only on the skip changes, as advertised in the title.

> but you can now run `./check.sh itest substring`

This syntax is used for listing the tasks that `check.sh` has to run: `test`, `itest`, `clippy`, `doc` etc. So it would be confusing to recycle it for filtering. An explicit `--filter` argument would be more intuitive (but let's discuss this in another PR).

---

In general, it would be nice if I didn't have to repeat comments I already gave multiple times before. One of them is the formatting of `quote! { ... }` macros, which I mentioned at least four times in your last PR ([1](https://github.com/godot-rust/gdext/pull/246#discussion_r1180080411) [2](https://github.com/godot-rust/gdext/pull/246#discussion_r1187824892) [3](https://github.com/godot-rust/gdext/pull/246#discussion_r1187830187) [4](https://github.com/godot-rust/gdext/pull/246#discussion_r1198906245)). Unfortunately, this PR suffers from the exact same issues.

Reviewing pull requests takes a lot of time during which I cannot work on implementation, so I would appreciate if I could spend this time on more meaningful issues. We try to automate whatever we can for bikeshedding topics like formatting, but rustfmt isn't perfect, and so I rely on user cooperation. I know this wasn't on purpose, but please write those review comments down, use the "mark as resolved" feature or any other system that works for you. Thanks a lot for your understanding!

astrale-sharp commented:
> In general, it would be nice if I didn't have to repeat comments I already gave multiple times before. One of them is the formatting of quote! { ... } macros, which I mentioned at least four times in your last PR (https://github.com/godot-rust/gdext/pull/246#discussion_r1180080411 https://github.com/godot-rust/gdext/pull/246#discussion_r1187824892 https://github.com/godot-rust/gdext/pull/246#discussion_r1187830187 https://github.com/godot-rust/gdext/pull/246#discussion_r1198906245). Unfortunately, this PR suffers from the exact same issues.

> Reviewing pull requests takes a lot of time during which I cannot work on implementation, so I would appreciate if I could spend this time on more meaningful issues. We try to automate whatever we can for bikeshedding topics like formatting, but rustfmt isn't perfect, and so I rely on user cooperation. I know this wasn't on purpose, but please write those review comments down, use the "mark as resolved" feature or any other system that works for you. Thanks a lot for your understanding!

Hi!

First thank you so much for all the reviewing, explaining, communicating, I am conscious it's a lot of work.

about the formatting, as long as I don't go out of draft state and or request a review it's that it's not ready and you're likely the waste your time if you review the whole thing.

I still like to create a draft to (order of decreasing importance)
- show that I'm working on something 
- have a place to ask specific questions about the draft
- make the process visible for whoever is interested
- I like github's diff viewer

I have two solution in mind if this is not working for you: (order of decreasing preference)
- I open a draft so I show that i'm working on something but I only push when it's ready for review.
- I don't open a draft and only create PR's when they're ready for review

Do one of these solutions work for you? or maybe you have something else in mind?

Sorry if I made you lose time! Really appreciate you and your work!

astrale-sharp commented:
notify uses a rust version that is too old for compiling the PR, that's why it fails.

Bromeon commented:
> about the formatting, as long as I don't go out of draft state and or request a review it's that it's not ready and you're likely the waste your time if you review the whole thing.
> 
> I still like to create a draft to (order of decreasing importance)
> 
>   * show that I'm working on something
>   * have a place to ask specific questions about the draft
>   * make the process visible for whoever is interested
>   * I like github's diff viewer
> 
> 
> I have two solution in mind if this is not working for you: (order of decreasing preference)
> 
>   * I open a draft so I show that i'm working on something but I only push when it's ready for review.
>   * I don't open a draft and only create PR's when they're ready for review
> 
> 
> Do one of these solutions work for you? or maybe you have something else in mind?

Thank you for elaborating your workflow! That explains a lot.

Normally I ignore draft pushes, but in this case you wrote a message, which I interpreted as "here's what I did, please have a look". Sorry for the misunderstanding.

One big problem for me is that pushes to draft PRs create notifications. This is a [known issue](https://github.com/orgs/community/discussions/10409), but GitHub has been ignoring it for almost 1.5 years, so no hope that anything will change in the near future. Since I don't know whether there are other important notifications, I still need to check the bell every now and then.

---


For the diff viewer, an option would be to open a PR within your own fork, but that defeats the purpose of showing/asking about ongoing work to the community.

A better idea: I can unsubscribe from your PRs if they are opened as drafts -- then you would need to mention me @bromeon once you're ready. I'm not sure if/how well that works in practice, but it's something we could try?

astrale-sharp commented:
My preferred solution is if you unsubscribe from my draft and then I @ you when I'm ready!

astrale-sharp commented:
after overriding my version to 1.66.0, i get clippy errors from changes not related to mine, should I fix them? (2 very easy fix)
```
error: use of `unwrap_or` followed by a function call
   --> itest/rust/src/virtual_methods_test.rs:538:14
    |
538 |             .unwrap_or(Variant::nil())
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(Variant::nil)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
    = note: `-D clippy::or-fun-call` implied by `-D clippy::perf`

error: use of `unwrap_or` followed by a call to `new`
   --> itest/rust/src/runner.rs:248:10
    |
248 |         .unwrap_or(Array::new())
    |          ^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_default()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

astrale-sharp commented:
In the future I'd like to add support for a 
#[variant( skip (default = user_func()))]
Should I add it in this pr ?

if so we'd need to discuss the syntax, a few come to mind ( I'll copy this message elsewhere if you'd rather have a separate pr)
```rust
struct UserStruct {
        #[variant( skip = (default = %value% ) )]
        field : ty
        #[to_variant( skip = (default = %value% ) )] // same but only skip 
}
```
I feel like there is a better way to do this

Bromeon commented:
> In the future I'd like to add support for a
> #[variant( skip (default = user_func()))]
> Should I add it in this pr ?

No, let's discuss separately. This PR is already open for 1.5 months, I'd like to wrap it up asap...

astrale-sharp commented:
Then I believe there's only `roundtrip_with_skip` to decide and then  the pr should be good !

Bromeon commented on itest/rust/src/derive_variant/structs_skip.rs:
```diff
@@ -0,0 +1,62 @@
+
+#[itest]
+fn new_type_to_variant() {
+    let s = NewTypeStructWithSkip("four".into());
```
Prefer `to_string()` over `into()` for all these string conversions.

Bromeon commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+                quote!{ #ident :#ident.try_to()? }
+            ),
+        }
+
+
```
```suggestion
```

Bromeon commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+        #(
+            #set_idents
+        )*
+        Ok(Self{ #(#set_self,)* })
```
```suggestion
        Ok(Self { #(#set_self,)* })
```

Bromeon commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+    name: &impl ToTokens,
+) {
+    let ident_and_set = fields.fields.iter().enumerate()
+        .map(|(k, (f,_))| {
```
```suggestion
        .map(|(k, (f, _))| {
```

Bromeon commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+    *body = quote! {
+        #body
+        let mut root = root.try_to::<godot::builtin::VariantArray>()?;
+        #(
+            #ident_set
+
+        )*
```
```suggestion
        #(
            #ident_set
        )*
```

Bromeon commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+                let #ident = <#field_type as Default>::default();
+            },
+            false => quote! {
+                    let #ident = variant.pop_front()
+                                    .ok_or(godot::builtin::VariantConversionError::MissingValue)?
+                                    .try_to::<#field_type>()?;
```
indentation

Bromeon commented on itest/rust/src/derive_variant/mod.rs:
```diff
@@ -9,7 +9,42 @@ use std::fmt::Debug;
+
+fn roundtrip<T, U>(value: T, expected: U)
+where
+    T: ToVariant + FromVariant + std::cmp::PartialEq + Debug,
```
`PartialEq` is in the Rust prelude, no need to qualify it.

Bromeon commented on itest/rust/src/derive_variant/mod.rs:
```diff
@@ -9,7 +9,42 @@ use std::fmt::Debug;
+//#[macro_use]
+mod structs_skip;
+
+fn roundtrip<T, U>(value: T, expected: U)
+where
+    T: ToVariant + FromVariant + std::cmp::PartialEq + Debug,
+    U: ToVariant,
+{
+    let expected = expected.to_variant();
+
+    assert_eq!(value.to_variant(), expected, "testing converting to");
+    assert_eq!(
+        value,
+        T::from_variant(&expected),
+        "testing converting back from"
+    );
+}
```
This exact function already exists in `variant_test.rs`. Consider moving it one level up to `itest/src/lib.rs` and reuse it.

astrale-sharp replied:
it uses only one argument which offers less assurance in our case but it's still a pretty good test so i changed it for now but didn't squash the commit yet in case it's better with two arguments

Bromeon reviewed (changes requested):
Unfortunately there are still tons of formatting issues. Please, _please_ go through all your code and make sure macros are consistently formatted. In case of doubt, **look at existing code**. I'm sorry to be pedantic here, but we have to maintain the quality of code for readers and the alternative is that I have to fix all those little things myself. I'm also really disappointed in rustfmt to not catch any of these; might consider using an alternative formatter in the future...

Don't use `match` if you only have `true/false` branches; this is what `if` is for.

The last thing: it would be nice if this macro got some documentation, otherwise you spent all that effort any people may not even discover this API. https://godot-rust.github.io/docs/gdext/pr-251/godot/bind/derive.ToVariant.html is currently empty. No need to write a novel, but maybe quickly elaborate the purpose and include 3-4 examples as doctests (you can even take those from the tests). Same for `FromVariant`. Also explain the `#[skip]` in particular.

astrale-sharp commented on godot-macros/src/derive_from_variant.rs:
```diff
@@ -198,3 +104,199 @@ pub fn transform(decl: Declaration) -> ParseResult<TokenStream> {
+                    let #ident = root.get(#string_ident).ok_or(godot::builtin::VariantConversionError::MissingValue)?;
+                },
+
+                quote!{ #ident :#ident.try_to()? }
```
Running code inside quote! through rustfmt very generally doesn't work.

Now, I have a handicap which makes it difficult to read correctly, especially formatting nitpicks, I'm going to try to fix these by hand and then I probably won't contribute macro code anymore, I have the feeling that I make you lose time more than anything else.

I have tried using rustfmt with these options : format_macro_matchers=true,format_macro_bodies=true, error_on_unformatted=true but unfortunatly they don't seem to work for proc macros.

astrale-sharp replied:
error on unformatted might still be of interest to you though

Bromeon replied:
Sorry to hear that. One option we could try is https://github.com/andrewbaxter/genemichaels -- its mission statement is that it formats macros. Maybe that automation could help!

astrale-sharp replied:
I'll definitely try it ! really happy to hear such a thing exists!

astrale-sharp replied:
I tried it and the result is absolutely awful...

astrale-sharp replied:
I've used this regex `(?<!:)(:\w)` to check the code for this

astrale-sharp commented on itest/rust/src/variant_test.rs:
```diff
@@ -389,7 +389,7 @@ fn variant_hash_correct() {
         varray![false, true, 4, "7"].to_variant(),
         0.to_variant(),
-        dict! { 0 : dict!{ 0: 1 }}.to_variant(),
+        dict! { 0 : dict!{ 0: 1 } }.to_variant(),
```
caught this with regexes

astrale-sharp commented:
I used a set of regexes to help me, hoping i caught all the formatting problems.

I don't have much motivation to keep working on this.

Bromeon reviewed (commented):
> I don't have much motivation to keep working on this.

Fully understandable. Maybe some tips for the future:
* Split the PR into smaller parts. This turned out to be quite an involved feature, but you could for example have started only with `#[variant(skip)]` for structs. This would reduce the amount of code to review, the amount of changes to address, and the overall work involved. Don't think you are expected to do "the full thing", a lot of additions are helpful in a partial state and can be extended later, by you or someone else.
* Motivation is generally higher if the time between iterations is shorter. After several months, a lot of previous points are forgotten, it needs more time to re-think into the problem, etc. I understand there may be all sorts of RL reasons that delay pull requests, but having a smaller PR (point 1) would have the biggest impact here.
* Regarding the formatting stuff, I don't like to nitpick a lot, but at the same time please consider that work that is not done during the PR likely has to be done at some point in the future, and likely by me 😬 so I'd like to keep a level of code quality, if only because it helps other contributors navigate and recognize a consistent style. I hope that makes sense.

Thank you very much for your hard work on this feature addition, as well as the patience and persistence! 👍 

bors r+
